core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -2904,6 +2907,33 @@ public static boolean classifyFilters(
     return !filtersToRemove.isEmpty();
   }
 
+  private static final RelVisitor LIMIT_FINDER = new RelVisitor() {
+    @Override public void visit(RelNode node, int ordinal, @Nullable RelNode parent) {
+      if (isLimit(node)) {
+        throw Util.FoundOne.NULL;
+      }
+
+      if (node instanceof HepRelVertex) {
```

Comment:
Can we avoid using HepRelVertex? I am afraid it might be planner specific. Have you tried to run the query on Volcano planner?

Reply:
@hsyuan I'm already reviewing in JIRA, and had made similar comments there.

Reply:
Ah, sorry for not noticing the discussion in JIRA. will move there.